src/Neo/Extensions/Persistence/ReadOnlyViewExtensions.cs

```diff
@@ -0,0 +1,157 @@
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static IEnumerable<(StorageKey Key, StorageItem Value)> ScanPrefix(
+            this IReadOnlyStoreView view,
+            byte[]? keyPrefix,
+            SeekDirection direction = SeekDirection.Forward)
+        {
+            var seekPrefix = direction == SeekDirection.Forward ? keyPrefix : keyPrefix.GetSeekPrefix();
+            return view.ScanPrefix(keyPrefix, seekPrefix, direction);
+        }
+
+        internal static IEnumerable<(StorageKey Key, StorageItem Value)> ScanPrefix(
+            this IReadOnlyStoreView view,
+            byte[]? keyPrefix,
+            byte[]? seekPrefix,
+            SeekDirection direction = SeekDirection.Forward)
+        {
+            foreach (var (key, value) in view.Seek(seekPrefix, direction))
+            {
+                if (keyPrefix == null || key.ToArray().AsSpan().StartsWith(keyPrefix))
+                    yield return new(key, value);
+                else if (direction == SeekDirection.Forward || (seekPrefix == null || !key.ToArray().SequenceEqual(seekPrefix)))
+                    yield break;
+            }
+        }
+
+        /// <summary>
+        /// Scans the entries in the specified range.
+        /// <para>
+        /// If <paramref name="direction"/> is <see cref="SeekDirection.Forward"/>,
+        /// it seeks to the first entry if <paramref name="inclusiveStartKey"/> is null or empty.
+        /// </para>
+        /// <para>
+        /// If <paramref name="direction"/> is <see cref="SeekDirection.Backward"/>,
+        /// the <paramref name="inclusiveStartKey"/> cannot be null or empty.
+        /// </para>
+        /// <para>
+        /// If want to scan all entries with <see cref="SeekDirection.Backward"/>,
+        /// set <paramref name="inclusiveStartKey"/> to be N * 0xff and <paramref name="exclusiveEndKey"/> to be empty,
+        /// where N is the max length of the key.
+        /// </para>
+        /// </summary>
+        /// <param name="view">The view to scan.</param>
+        /// <param name="inclusiveStartKey">The inclusive start key.</param>
+        /// <param name="exclusiveEndKey">The exclusive end key.</param>
+        /// <param name="direction">The search direction.</param>
+        /// <returns>The entries found in the specified range.</returns>
+        public static IEnumerable<(StorageKey Key, StorageItem Value)> ScanRange(
+            this IReadOnlyStoreView view,
+            byte[]? inclusiveStartKey,
+            byte[] exclusiveEndKey,
+            SeekDirection direction = SeekDirection.Forward)
+        {
+            ByteArrayComparer comparer = direction == SeekDirection.Forward
+                ? ByteArrayComparer.Default
+                : ByteArrayComparer.Reverse;
+            foreach (var (key, value) in view.Seek(inclusiveStartKey, direction))
+            {
+                if (comparer.Compare(key.ToArray(), exclusiveEndKey) < 0)
```

Comment:
You want the the key. If the prefix is a key?
```suggestion
                if (comparer.Compare(key.ToArray(), exclusiveEndKey) <= 0)
```

Reply:
> You want the the key. If the prefix is a key?

`exclusiveEndKey`

because of exclusive

Reply:
Why go to all the trouble to edit or change this. If your not going to fix the issues with the class.

Reply:
> Why go to all the trouble to edit or change this. If your not going to fix the issues with the class?

Becase this is `ScanRange`/`FindRange`, and range is [StartKey, EndKey)

Reply:
it doesnt matter. In the stores the key are in order. If you convert all keys to an `integer` they start from `zero through Infinity` (_or biggest key_). so if last key is `[0xff, 0xff]` and i use prefix `0xff` the seeker goto `0xff` key.

Example with prefix `0xff` for keys ranging from `[0x01]` to `[0xff, 0xff]`

Records returned in order:
```
Key: [0xff]
Key: [0xfe]
Key: [0xfd]
Key: [0xfc]
Key: [0xfb]
Key: [0xfa]
.... etc ...
Key: [0xdf]
Key: [0xde]
Key: [0xdd]
Key: [0xdc]
Key: [0xdb]
Key: [0xda]
.... etc ...
Key: [0x01]
```

Reply:
> prefix is misleading word.
> 
> it doesnt matter. In the stores the key are in order. If you convert all keys to an `integer` they start from `zero through Infinity` (_or biggest key in the store_). so if last key is `[0xff, 0xff]` and i use prefix `0xff` the seeker goto `0xff` key.
> 
> Example with prefix `0xff` for keys ranging from `[0x01]` to `[0xff, 0xff]`
> 
> Records returned in order:
> 
> ```
> Key: [0xff]
> Key: [0xfe]
> Key: [0xfd]
> Key: [0xfc]
> Key: [0xfb]
> Key: [0xfa]
> .... etc ...
> Key: [0xdf]
> Key: [0xde]
> Key: [0xdd]
> Key: [0xdc]
> Key: [0xdb]
> Key: [0xda]
> .... etc ...
> Key: [0x01]
> ```

`ScanRange`/ `FindRange` return items within range [StartKey(inclusive), EndKey(exclusive) )

Reply:
Where are you getting this information? What is not to understand? `IStore` doesn't work this way. `Find` has bugs in `DataCache`.

Reply:
> Where are you getting this information? What is not to understand? `IStore` doesn't work this way. `Find` has bugs in `DataCache`.

https://github.com/neo-project/neo/blob/master/src/Neo/Persistence/DataCache.cs#L264

Reply:
Like I said, `DataCache` has bugs.

#### LevelDBStore

https://github.com/neo-project/neo/blob/84a9fd5702f9b64dfac519b75243d7790c63bc4c/src/Plugins/LevelDBStore/IO/Data/LevelDB/Helper.cs#L21-L43

#### RocksDbStore

https://github.com/neo-project/neo/blob/84a9fd5702f9b64dfac519b75243d7790c63bc4c/src/Plugins/RocksDBStore/Plugins/Storage/Store.cs#L41-L52

#### MemoryStore

https://github.com/neo-project/neo/blob/84a9fd5702f9b64dfac519b75243d7790c63bc4c/src/Neo/Persistence/MemoryStore.cs#L51-L63

Reply:
> Like I said, `DataCache` has bugs.
> 
> #### LevelDBStore
> https://github.com/neo-project/neo/blob/84a9fd5702f9b64dfac519b75243d7790c63bc4c/src/Plugins/LevelDBStore/IO/Data/LevelDB/Helper.cs#L21-L43
> 
> #### RocksDbStore
> https://github.com/neo-project/neo/blob/84a9fd5702f9b64dfac519b75243d7790c63bc4c/src/Plugins/RocksDBStore/Plugins/Storage/Store.cs#L41-L52
> 
> #### MemoryStore
> https://github.com/neo-project/neo/blob/84a9fd5702f9b64dfac519b75243d7790c63bc4c/src/Neo/Persistence/MemoryStore.cs#L51-L63

If `DataCache` has bugs, you can submit a PR to fix it.

Reply:
This is not `**Find**`.

`ScanRange` /  `FindRange` return items within range [StartKey(inclusive), EndKey(exclusive) )

Reply:
> prefix is misleading word.
> 
> it doesnt matter. In the stores the key are in order. If you convert all keys to an `integer` they start from `zero through Infinity` (_or biggest key in the store_). so if last key is `[0xff, 0xff]` and i use prefix `0xff` the seeker goto `0xff` key.
> 
> Example with prefix `0xff` for keys ranging from `[0x01]` to `[0xff, 0xff]`
> 
> Records returned in order:
> 
> ```
> Key: [0xff]
> Key: [0xfe]
> Key: [0xfd]
> Key: [0xfc]
> Key: [0xfb]
> Key: [0xfa]
> .... etc ...
> Key: [0xdf]
> Key: [0xde]
> Key: [0xdd]
> Key: [0xdc]
> Key: [0xdb]
> Key: [0xda]
> .... etc ...
> Key: [0x01]
> ```

You can read this:
https://github.com/neo-project/neo/blob/84a9fd5702f9b64dfac519b75243d7790c63bc4c/src/Neo/Persistence/DataCache.cs#L251-L258